crates/circuit/src/parameter_table.rs

```diff
@@ -10,164 +10,395 @@
 // copyright notice, and modified files need to carry a notice indicating
 // that they have been altered from the originals.
 
+use hashbrown::hash_map::Entry;
+use hashbrown::{HashMap, HashSet};
+use thiserror::Error;
+
+use pyo3::exceptions::PyTypeError;
 use pyo3::prelude::*;
-use pyo3::{import_exception, intern, PyObject};
+use pyo3::pybacked::PyBackedStr;
+use pyo3::types::{PyList, PySet};
+use pyo3::{import_exception, intern, PyTraverseError, PyVisit};
 
-import_exception!(qiskit.circuit.exceptions, CircuitError);
+use crate::imports::UUID;
 
-use hashbrown::{HashMap, HashSet};
+import_exception!(qiskit.circuit, CircuitError);
+
+#[derive(Error, Debug)]
+pub enum ParameterTableError {
+    #[error("parameter '{0:?}' is not tracked in the table")]
+    ParameterNotTracked(ParameterUuid),
+    #[error("usage {0:?} is not tracked by the table")]
+    UsageNotTracked(ParameterUse),
+}
+impl From<ParameterTableError> for PyErr {
+    fn from(value: ParameterTableError) -> PyErr {
+        CircuitError::new_err(value.to_string())
+    }
+}
+
+/// A single use of a symbolic parameter.
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+pub enum ParameterUse {
+    Index { instruction: usize, parameter: u32 },
```

Comment:
`instruction` indexes into `CircuitData::data`, which has no bounds on it.  `parameter` indexes into an instruction's `params` field, but the `Operation` trait we have in Rust space requires `num_params` to be a `u32`, so a `u32` is sufficient to index that.

On a 64-bit system, this wouldn't actually cause any different in size of the `Index` variant, because `usize` is 8 bytes and `u32` is 4, so alignment of the struct would require that it gets padded up to 16 bytes.  _However_, this is defined inline in an `enum`, so we also need to store the enum discriminant in that data as well, which is an extra bit of data.  By ensuring that there's at least a padding byte in all variants, we can enable a Rust compiler optimisation that stores the variant in what would otherwise be padding, so `ParameterUse` remains 16 bytes, rather than naively getting expanded to 24.

On 32-bit systems, `ParameterUse` is 12 bytes and we don't get the padding trick, but we don't care much about that.

Reply:
Nice, thanks for explanation 😄